connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -409,11 +405,26 @@ class SparkSession private[sql] (
     val plan = proto.Plan.newBuilder().setCommand(command).build()
     // .toSeq forces that the iterator is consumed and closed. On top, ignore all
     // progress messages.
-    client.execute(plan).filter(!_.hasExecutionProgress).toSeq
+    execute(plan).filter(!_.hasExecutionProgress).toSeq
   }
 
-  private[sql] def execute(plan: proto.Plan): CloseableIterator[ExecutePlanResponse] =
-    client.execute(plan)
+  /**
+   * The real `execute` method that calls into `SparkConnectClient`.
+   *
+   * Here we inject a lazy map to process registered observed metrics, so consumers of the returned
+   * iterator does not need to worry about it.
+   *
+   * Please make sure all `execute` overrides call this method.
+   */
+  private[sql] def execute(plan: proto.Plan): CloseableIterator[ExecutePlanResponse] = {
```

Comment:
Would it make sense to rename this to `executeInternal` so it is clear that all the overrides are calling this specific method?

Reply:
Good idea. Done!